[KEP-5365] Implement Image Volume with Digest

iholder101 opened the pull request:
/kind feature
/sig node

#### What type of PR is this?

#### What this PR does / why we need it:

#### Which issue(s) this PR is related to:
KEP: https://github.com/kubernetes/enhancements/issues/5365

#### Does this PR introduce a user-facing change?
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
The ImageVolumeWithDigest is added which adds the digest of image volumes to the container's status.
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:
```docs
- KEP: https://github.com/kubernetes/enhancements/blob/master/keps/sig-node/5365-ImageVolume-with-image-digest/README.md
```

iholder101 committed:
cri-api: Add image_ref to the ImageSpec struct

Signed-off-by: Itamar Holder <iholder@redhat.com>

iholder101 committed:
refactor: expose imageSpec conversion functions

Signed-off-by: Itamar Holder <iholder@redhat.com>

iholder101 committed:
Set image volume digest as part of container statuses

Signed-off-by: Itamar Holder <iholder@redhat.com>

iholder101 committed:
Add ImageRef field to VolumeMountStatus

Signed-off-by: Itamar Holder <iholder@redhat.com>

iholder101 committed:
add the ImageVolumeWithDigest feature gate

Signed-off-by: Itamar Holder <iholder@redhat.com>

haircommander commented on pkg/kubelet/kubelet_pods.go:
```diff
@@ -1993,6 +1994,16 @@ func (kl *Kubelet) convertStatusToAPIStatus(pod *v1.Pod, podStatus *kubecontaine
 // statuses into API container statuses.
 func (kl *Kubelet) convertToAPIContainerStatuses(pod *v1.Pod, podStatus *kubecontainer.PodStatus, previousStatus []v1.ContainerStatus, containers []v1.Container, hasInitContainers, isInitContainer bool) []v1.ContainerStatus {
+	imageVolumeNames := map[string]struct{}{}
+	if utilfeature.DefaultFeatureGate.Enabled(features.ImageVolume) && utilfeature.DefaultFeatureGate.Enabled(features.ImageVolumeWithDigest) {
```
should we cache this for the next call?

iholder101 commented:
I'm not entirely sure what this failure means, but I guess that I've screwed things up with the conversion function (see [this commit](https://github.com/kubernetes/kubernetes/pull/132807/commits/59642910e5713b2c95c5650a0315cb23b04b7163))?

```
{Failed  === RUN   TestCompatibility/core.v1.Pod/HEAD
    compatibility.go:248: json differs
    compatibility.go:249:   (
          	"""
          	... // 1824 identical lines
          	            "mountPath": "mountPathValue",
          	            "readOnly": true,
        - 	            "recursiveReadOnly": "recursiveReadOnlyValue"
        + 	            "recursiveReadOnly": "recursiveReadOnlyValue",
        + 	            "imageRef": "imageRefValue"
          	          }
          	        ],
          	... // 88 identical lines
          	            "mountPath": "mountPathValue",
          	            "readOnly": true,
        - 	            "recursiveReadOnly": "recursiveReadOnlyValue"
        + 	            "recursiveReadOnly": "recursiveReadOnlyValue",
        + 	            "imageRef": "imageRefValue"
          	          }
          	        ],
          	... // 89 identical lines
          	            "mountPath": "mountPathValue",
          	            "readOnly": true,
        - 	            "recursiveReadOnly": "recursiveReadOnlyValue"
        + 	            "recursiveReadOnly": "recursiveReadOnlyValue",
        + 	            "imageRef": "imageRefValue"
          	          }
          	        ],
          	... // 32 identical lines
          	"""
          )
```

Would appreciate some guidance on that

iholder101 committed:
Add imageRef to testdata

by running:
UPDATE_COMPATIBILITY_FIXTURE_DATA=true make test WHAT=./staging/src/k8s.io/api

Signed-off-by: Itamar Holder <iholder@redhat.com>

iholder101 committed:
unit test: Set image volume digest as part of container statuses

Signed-off-by: Itamar Holder <iholder@redhat.com>

iholder101 commented:
/remove-sig api-machinery

haircommander commented:
/lgtm

kannon92 commented on staging/src/k8s.io/api/core/v1/types.go:
```diff
@@ -3605,6 +3605,10 @@ type VolumeMountStatus struct {
 	// depending on the mount result.
 	// +optional
 	RecursiveReadOnly *RecursiveReadOnlyMode `json:"recursiveReadOnly,omitempty" protobuf:"bytes,4,opt,name=recursiveReadOnly,casttype=RecursiveReadOnlyMode"`
+	// ImageRef is the digest of the image used for this volume.
```
```suggestion
	// imageRef is the digest of the image used for this volume.
```

thockin replied:
We are wildly inconsistent about this

iholder101 commented on staging/src/k8s.io/api/core/v1/types.go:
```diff
@@ -3605,6 +3605,10 @@ type VolumeMountStatus struct {
+	// ImageRef is the digest of the image used for this volume.
+	// It should have a value that's similar to the pod's status.containerStatuses[i].imageID.
+	// If the volume source is not an ImageVolume, this field will be empty.
+	ImageRef *string `json:"imageRef,omitempty" protobuf:"bytes,5,opt,name=imageRef"`
```
Thanks for the input @thockin. Do you have an alternative home for this field in mind?

cc @haircommander @deads2k

iholder101 replied:
Done! @thockin 
Thanks for raising that

SergeyKanzhelev commented on pkg/kubelet/kubelet_pods.go:
```diff
@@ -2130,6 +2145,39 @@ func (kl *Kubelet) convertToAPIContainerStatuses(pod *v1.Pod, podStatus *kubecon
+					klog.ErrorS(err, "error getting image volume digest", "image name", imageVolumeNames[volumeName])
+				}
+
+				status.VolumeMounts[i].ImageRef = &imageRef
```
should this be under the else? If err != nil, `imageRef` will not be defined, but we still assign the value

SergeyKanzhelev commented on pkg/kubelet/kubelet_pods.go:
```diff
@@ -2130,6 +2145,39 @@ func (kl *Kubelet) convertToAPIContainerStatuses(pod *v1.Pod, podStatus *kubecon
+					continue
+				}
+
+				imageRef, err := kl.containerRuntime.GetImageRef(context.Background(), imageSpec)
```
if you want to validate some of this logic, especially error cases, it can be done via criproxy tests

thockin commented on pkg/apis/core/types.go:
```diff
@@ -3183,6 +3183,10 @@ type VolumeMountStatus struct {
+	// ImageRef is the digest of the image used for this volume.
+	// It should have a value that's similar to the pod's status.containerStatuses[i].imageID.
+	// If the volume source is not an ImageVolume, this field will be empty.
+	ImageRef *string
```
You want this to match the v1 structure -- you should not need to hand-write conversion code.

thockin reviewed (commented):
This PR is missing:

1) field-wiping logic - you need to add this to `dropDisabledPodStatusFields`.  If the gate is not set AND this was not already set in oldPod, then wipe it.
2) validation - you need to check that exactly one field in `ImageVolumeStatus` is set, but only if the gate is enabled.  But we don't check gates in validation, we check them in strategy (see above) and set a (new) field in opts.  Validation checks the opts.

In short, if the gate is enabled, we validate.  If someone is already using this field (e.g. in an update), we treat it as if the gate was enabled.  If they are not using it already and the gate is off, we treat it as if the field does not exist (wipe).

iholder101 commented on staging/src/k8s.io/api/core/v1/types.go:
```diff
@@ -3605,6 +3605,10 @@ type VolumeMountStatus struct {
 	// depending on the mount result.
 	// +optional
 	RecursiveReadOnly *RecursiveReadOnlyMode `json:"recursiveReadOnly,omitempty" protobuf:"bytes,4,opt,name=recursiveReadOnly,casttype=RecursiveReadOnlyMode"`
+	// ImageRef is the digest of the image used for this volume.
```
I've added some fields.
It's the first time I'm using CRIProxy, hope it makes sense :+1: let me know

iholder101 committed:
Add field wiping and validation logic

Signed-off-by: Itamar Holder <iholder@redhat.com>

thockin commented on pkg/apis/core/validation/validation.go:
```diff
@@ -5930,6 +5932,27 @@ func ValidatePodStatusUpdate(newPod, oldPod *core.Pod, opts PodValidationOptions
+	if opts.AllowImageVolumeWithDigest {
+		for _, containerStatus := range newPod.Status.ContainerStatuses {
+			for _, volumeMountStatus := range containerStatus.VolumeMounts {
+				validatedField := fldPath.Child("containerStatuses", "volumeMounts", "volumeStatus")
```
The correct fieldpath would be something like `fldPath.Child("containerStatuses").Index(i).Child("volumeMounts").Index(j).Child("volumeStatus")`

thockin commented on staging/src/k8s.io/api/core/v1/types.go:
```diff
@@ -3605,6 +3605,26 @@ type VolumeMountStatus struct {
+// ImageVolumeStatus represents the image-based volume status.
+type ImageVolumeStatus struct {
+	// ImageRef is the digest of the image used for this volume.
+	// It should have a value that's similar to the pod's status.containerStatuses[i].imageID.
```
Do we really have no spec for it?

Is `""` a valid value?  Is `"x"` valid?  Is `"Four score and seven years ago, ..." valid?  I suspect you want to spec it as non-empty and limit the length to something like 128 bytes?

SergeyKanzhelev replied:
historically we have issues with what is expected here: https://github.com/kubernetes/kubernetes/issues/115199

We cannot enforce specific value for this field in this KEP, but since KEP is using the same CRI APIs to pull images and check status.

I think this inconsistency can and should be fixed in future updates to ImageService CRI APIs

iholder101 replied:
> Can we just assert that it can't be empty and it must be less than some number (e.g. 256, 512, 1024) bytes?

This makes sense to me.
Hope this is fine with you as well @SergeyKanzhelev.
Done.

iholder101 committed:
Add auto-generated code and boilerplate

Signed-off-by: Itamar Holder <iholder@redhat.com>

iholder101 committed:
Add e2e tests

Signed-off-by: Itamar Holder <iholder@redhat.com>

haircommander commented:
/retest

SergeyKanzhelev commented on pkg/api/pod/util.go:
```diff
@@ -1054,6 +1055,13 @@ func dropDisabledPodStatusFields(podStatus, oldPodStatus *api.PodStatus, podSpec
 	}
+
+	imageVolumeDisabled := !utilfeature.DefaultFeatureGate.Enabled(features.ImageVolume)
+	imageVolumeDigestDisabled := imageVolumeDisabled || !utilfeature.DefaultFeatureGate.Enabled(features.ImageVolumeWithDigest)
```
can be simplified since the dependency between feature gates is declared

SergeyKanzhelev commented on pkg/apis/core/validation/validation.go:
```diff
@@ -5930,6 +5932,27 @@ func ValidatePodStatusUpdate(newPod, oldPod *core.Pod, opts PodValidationOptions
+			}
+		}
+	}
+
```
checking for not being empty sounds like the only reasonable thing we can enforce. Do we enforce it for ImageID?

thockin replied:
Yeah, at the very least vaslidate that it is not `""`.  Would be good to validate that `len()` is less than some reasonable limit.  We can always bump the limit, with some effort, but unchecked values are a liability.

SergeyKanzhelev commented on test/e2e_node/criproxy_test.go:
```diff
@@ -94,6 +98,133 @@ var _ = SIGDescribe(feature.CriProxy, framework.WithSerial(), func() {
 
+	ginkgo.Context("Image volume digest error handling", func() {
+		ginkgo.BeforeEach(func() {
+			e2eskipper.SkipUnlessFeatureGateEnabled(features.ImageVolume)
```
are you doing this instead of adding WithFeatureGate to the test name because you want to execute it in some special test suite? I think it is best to follow the existing pattern

SergeyKanzhelev reviewed (commented):
logic looks lgtm. There are questions from @thockin but I do not thing that this KEP is the right one to answer some of them.

W.r.t. some next steps:

- Tests tags in e2e_node needs to be cleaned up. New tests needs a Feature.criproxy and a special FeatureGate. So it will need to have a new test lane in test-infra. This needs to be cleaned up with proper additional comments. I am OK to take it as a follow up.
- e2e tests will need to be added (under test/e2e) so we can eventually promote them to Conformance. This is GA graduation criteria, but starting early is always the best. Consider adding them if not in this release, than the next one

iholder101 commented on pkg/apis/core/validation/validation.go:
```diff
@@ -9318,3 +9341,31 @@ func validateNodeSwapStatus(nodeSwapStatus *core.NodeSwapStatus, fldPath *field.
 }
+
+// validateMaxOneFieldSet returns an error if the given struct has more than 1 populated field.
+func validateMaxOneFieldSet(v any, fldPath *field.Path) field.ErrorList {
```
Done.

iholder101 commented on staging/src/k8s.io/api/core/v1/types.go:
```diff
@@ -3620,6 +3620,26 @@ type VolumeMountStatus struct {
+	// ImageRef is the digest of the image used for this volume.
+	// It should have a value that's similar to the pod's status.containerStatuses[i].imageID.
+	// +required
+	ImageRef string `json:"imageRef,omitempty" protobuf:"bytes,1,opt,name=imageRef"`
```
I don't see a reason why not :+1:

SergeyKanzhelev commented on pkg/apis/core/types.go:
```diff
@@ -3198,6 +3198,25 @@ type VolumeMountStatus struct {
 	RecursiveReadOnly *RecursiveReadOnlyMode
+	// volumeStatus represents volume-type-specific status about the mounted
+	// volume.
+	VolumeStatus
```
this file is not very consistent with marking things optional Perhaps this should be marked as such explicitly as other fields in this type

```suggestion
    // +ptional
	VolumeStatus
```

SergeyKanzhelev commented on pkg/apis/core/types.go:
```diff
@@ -3198,6 +3198,25 @@ type VolumeMountStatus struct {
+	VolumeStatus
+}
+
+// VolumeStatus Represents the status of a mounted volume.
```
```suggestion
// VolumeStatus represents the status of a mounted volume.
```

SergeyKanzhelev commented on test/e2e_node/criproxy_test.go:
```diff
@@ -94,6 +97,131 @@ var _ = SIGDescribe(feature.CriProxy, framework.WithSerial(), func() {
 
 	})
 
+	ginkgo.Context("Image volume digest error handling", f.WithFeatureGate(features.ImageVolumeWithDigest), func() {
```
```suggestion
	ginkgo.Context("Image volume digest error handling", f.WithFeatureGate(features.ImageVolumeWithDigest, f.WithFeature(feature.CriProxy)), func() {
```

SergeyKanzhelev commented:
/test pull-kubernetes-node-e2e-cri-proxy-all-alpha-serial

thockin commented on pkg/apis/core/validation/validation.go:
```diff
@@ -9410,3 +9433,41 @@ func validateWorkloadReference(workloadRef *core.WorkloadReference, fldPath *fie
+	}
+
+	if numStatuses > maxStatuses {
+		allErrors = append(allErrors, field.Forbidden(fldPath, "may not specify more than 1 volume type"))
```
You defined a const for max, and then embedded 1 in the error :)

I don't think you really need the const, personally.

thockin commented on pkg/apis/core/validation/validation.go:
```diff
@@ -9410,3 +9433,41 @@ func validateWorkloadReference(workloadRef *core.WorkloadReference, fldPath *fie
+	}
+
+	if imageVolumeStatus.ImageRef == "" {
+		allErrors = append(allErrors, field.Invalid(fldPath.Child("imageRef"), imageVolumeStatus.ImageRef, "imageRef must not be empty"))
```
This should use `field.Required()` to get consistent error messages

thockin commented on pkg/apis/core/validation/validation.go:
```diff
@@ -9410,3 +9433,41 @@ func validateWorkloadReference(workloadRef *core.WorkloadReference, fldPath *fie
+
+	const maxImageRefLen = 256
+	if len(imageVolumeStatus.ImageRef) > maxImageRefLen {
+		allErrors = append(allErrors, field.Invalid(fldPath.Child("imageRef"), imageVolumeStatus.ImageRef, fmt.Sprintf("imageRef must not have length bigger than %d", maxImageRefLen)))
```
`field.TooLong()` for the same reason

thockin reviewed (commented):
This LGTM but just tiny things

iholder101 commented:
> Is this on an exception or just waiting until the next train?

@thockin Your comments caught me right before I had to travel (a very long trip) to Kubecon. I've just returned home recently. Therefore I wasn't able to address comments in time. I was sure we've missed the train, unfortunately.

However, I feel like this is 99% ready. So if an exception is still an option (is it?) I'm more than willing to push this forward.

WDYT?

iholder101 commented:
/test pull-kubernetes-unit

SergeyKanzhelev commented:
> Can you merge this so it will all be ready in advance for the next one?

I think you meant approving, not merging. It will be merged once thaw will start (once approved).

However to keep the paperwork in check, can you please send the PR for milestone bump in enhancement repository?

/milestone clear

SergeyKanzhelev commented:
@iholder101 can you please send the PRs to k/e repo to update the milestone for this KEP? Since it is approved, we can take it super early in the release cycle

SergeyKanzhelev commented:
Version bump is approved

/lgtm
/approve

liggitt commented on staging/src/k8s.io/api/core/v1/types.go:
```diff
@@ -3623,6 +3623,29 @@ type VolumeMountStatus struct {
+	// volumeStatus represents volume-type-specific status about the mounted
+	// volume.
+	// +optional
+	VolumeStatus `json:"volumeStatus,omitempty" protobuf:"bytes,5,opt,name=volumeStatus"`
```
this needs to be a pointer, or we change the serialization of every existing pod that includes VolumeMountStatus (both in json/yaml and in proto)

found by https://github.com/kubernetes/kubernetes/pull/137831

cc @deads2k @BenTheElder

SergeyKanzhelev replied:
do we need +optional if is it a pointer?

BenTheElder replied:
Filed https://github.com/kubernetes/kubernetes/pull/137836
